File: slidingwindow/76-minimum-window-substring/MinimumWindowSubstring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() == 0 || t.length() == 0) return "";

        HashMap<Character, Integer> tCount = new HashMap<>();
        for (char c : t.toCharArray()) {
            tCount.put(c, tCount.getOrDefault(c, 0) + 1);
        }

        HashMap<Character, Integer> window = new HashMap<>();
        int required = tCount.size();
        int formed = 0;

        int l = 0, r = 0;
        int minLen = Integer.MAX_VALUE;
        int start = 0;

        while (r < s.length()) {
            char c = s.charAt(r);
            window.put(c, window.getOrDefault(c, 0) + 1);

            if (tCount.containsKey(c) && window.get(c).intValue() == tCount.get(c).intValue()) {
                formed++;
            }

            while (l <= r && formed == required) {
                if (r - l + 1 < minLen) {
                    minLen = r - l + 1;
                    start = l;
                }

                char leftChar = s.charAt(l);
                window.put(leftChar, window.get(leftChar) - 1);
                if (tCount.containsKey(leftChar) && window.get(leftChar) < tCount.get(leftChar)) {
                    formed--;
                }
                l++;
            }
            r++;
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start + minLen);
    }
}

File: slidingwindow/76-minimum-window-substring/MinimumWindowSubstring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""

        t_count = Counter(t)
        window = {}
        required = len(t_count)
        formed = 0

        l, r = 0, 0
        min_len = float("inf")
        min_window = (0, 0)

        while r < len(s):
            char = s[r]
            window[char] = window.get(char, 0) + 1

            if char in t_count and window[char] == t_count[char]:
                formed += 1

            while l <= r and formed == required:
                if r - l + 1 < min_len:
                    min_len = r - l + 1
                    min_window = (l, r)

                left_char = s[l]
                window[left_char] -= 1
                if left_char in t_count and window[left_char] < t_count[left_char]:
                    formed -= 1
                l += 1

            r += 1

        return "" if min_len == float("inf") else s[min_window[0]: min_window[1] + 1]